### ISSUE TITLE
Problem with thumbnail class on Mac OS

### ISSUE THREAD

**Issue by widya-saseno** [eclipse/gef-classic#4]
Hi Team,

I have a problem with [Thumbnail.java](https://github.com/eclipse/gef-legacy/blob/master/org.eclipse.draw2d/src/org/eclipse/draw2d/parts/Thumbnail.java) in org.eclipse.draw2d.parts. On Mac the thumbnail image will be "rendered" multiple times (on Windows and Linux this problem is not appear). I supposed the problem was because [thumbnailGC.drawImage - line 199](https://github.com/eclipse/gef-legacy/blob/14563a9e1f2af636a5364d195cf07dbff6f35fa6/org.eclipse.draw2d/src/org/eclipse/draw2d/parts/Thumbnail.java#L199) like described in [https://bugs.eclipse.org/bugs/show_bug.cgi?id=543796](https://bugs.eclipse.org/bugs/show_bug.cgi?id=543796). 

My question; is there any plan to fix this issue or maybe create a workaround? (I read in README.md that this project is a legacy project/component which still maintained but will not developed further)

**Comment by Phillipus** [comment]
@widya-saseno 

I forked GEF and Draw2D and created my own version:

https://github.com/archimatetool/archi/blob/master/org.eclipse.draw2d/src/org/eclipse/draw2d/parts/Thumbnail.java

**Comment by BeckerWdf** [comment]
Do we have a possibility to revive gef-legacy and provide bugfix-releases for it?

**Comment by Phillipus** [comment]
> Do we have a possibility to revive gef-legacy and provide bugfix-releases for it?

There is a plan to revive it. Keep an eye on https://accounts.eclipse.org/mailing-list/gef-dev

**Comment by vogella** [comment]
> Do we have a possibility to revive gef-legacy and provide bugfix-releases for it?

Yes, the project has been reactivated. Please provide PR for your fixes.

**Comment by Phillipus** [comment]
Lars, I have a fix for this issue:

https://github.com/archimatetool/archi/blob/master/org.eclipse.draw2d/src/org/eclipse/draw2d/parts/Thumbnail.java

 I'm not sure whether the fix is the best one, but I can submit a PR for discussion if required. It could be submitted as a draft PR if necessary.

**Comment by vogella** [comment]
Thanks, please send a PR so that @azoitl or others with GEF experience can have a look.

**Comment by Phillipus** [comment]
For reference, this is a discussion of the underlying cause in SWT:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=543796

**Comment by BeckerWdf** [comment]
> For reference, this is a discussion of the underlying cause in SWT:
> 
> https://bugs.eclipse.org/bugs/show_bug.cgi?id=543796

So is there a way to fix this in SWT?

**Comment by Phillipus** [comment]
> So is there a way to fix this in SWT?

I don't think the SWT contributors figured out a fix.

**Comment by vogella** [comment]
I talked to @nnemkin and he plans to have a look into https://bugs.eclipse.org/bugs/show_bug.cgi?id=543796. Would be nice to fix that inside SWT

**Comment by nnemkin** [comment]
I've left some comments on the issue [on Bugzilla](https://bugs.eclipse.org/bugs/show_bug.cgi?id=543796#c33).

TLDR is that I think caching an Image GC is a bit of API misuse. Since SWT doesn't have an explicit GC.flush() method to synchronize Image reading and writing, it's best to create an Image GC locally and release it before drawing an Image. Moreover, I suggest doing it on all platforms and not just as a workaround for macOS.

Are GEF image contexts heavy enough to require caching? Do you observe slowdowns with the workaround?

Another important thing: it seems that the only reason Thumbnail uses an intermediate Image is to prevent clip reset. Most graphics APIs don't allow enlarging the clip (it prevents internal optimizations). GC.setClip in SWT does allow this, which makes it difficult to implement and a great candidate for deprecation. Can't you disallow clip enlarging _by convention_ and avoid temporary Image altogether?

**Comment by Phillipus** [comment]
I think we could consider this one fixed as there have been no feedback to the contrary.
